File: bindings/java/samples/Zlink.Samples/src/main/java/systems/zlink/samples/SampleSupport.java
/* SPDX-License-Identifier: MPL-2.0 */

package systems.zlink.samples;

import systems.zlink.contracts.core.RoutingId;
import systems.zlink.contracts.eventing.MonitorEventType;
import systems.zlink.contracts.eventing.SocketMonitor;
import systems.zlink.contracts.messaging.Message;
import systems.zlink.contracts.messaging.Received;
import systems.zlink.contracts.service.spot.Actor;
import systems.zlink.contracts.service.spot.ActorJoinDecision;
import systems.zlink.contracts.service.spot.ActorLocation;
import systems.zlink.contracts.service.spot.ActorRef;
import systems.zlink.contracts.service.spot.Claim;
import systems.zlink.contracts.service.spot.ClaimRecvResult;
import systems.zlink.contracts.service.spot.Dispatch;
import systems.zlink.contracts.service.spot.MeshNode;
import systems.zlink.contracts.service.spot.OperationId;
import systems.zlink.contracts.service.spot.OperationKind;
import systems.zlink.contracts.service.spot.ReadyBatch;
import systems.zlink.contracts.service.spot.ReadyDomain;
import systems.zlink.contracts.service.spot.ReceiveBatch;
import systems.zlink.contracts.service.spot.ReceiveRecord;
import systems.zlink.contracts.service.spot.RecordKind;
import systems.zlink.contracts.service.spot.Spot;
import systems.zlink.contracts.service.spot.StreamSessionBinding;
import systems.zlink.contracts.service.spot.StreamSessionService;
import systems.zlink.contracts.sockets.RecvFlags;
import systems.zlink.contracts.sockets.SendFlags;
import systems.zlink.contracts.sockets.StreamSocket;
import systems.zlink.contracts.core.Zlink;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.InetSocketAddress;
import java.net.ServerSocket;
import java.time.Duration;
import java.util.List;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.function.Consumer;

final class SampleSupport {
    private static final Duration TIMEOUT = Duration.ofSeconds(5);
    static final String PAIR_PAYLOAD = "hello-pair";
    static final String DEALER_REQUEST = "ping";
    static final String DEALER_REPLY = "pong";
    static final String STREAM_PAYLOAD = "hello-stream";
    static final String PUBSUB_TOPIC = "prices";
    static final String PUBSUB_PAYLOAD = "101.25";
    private static final int TCP_PORT_BASE =
        40_000 + (int) (ProcessHandle.current().pid() % 10_000);
    private static final AtomicInteger NEXT_TCP_PORT =
        new AtomicInteger(TCP_PORT_BASE);

    private SampleSupport() {
    }

    /** Handles one received record during a {@link #pumpReady} drain. */
    @FunctionalInterface
    interface RecordHandler {
        void handle(ReceiveRecord record, ReceiveBatch batch, int index);
    }

    static String tcpEndpoint() {
        return "tcp://127.0.0.1:" + nextTcpPort();
    }

    private static int nextTcpPort() {
        for (int attempt = 0; attempt < 200; attempt++) {
            int port = NEXT_TCP_PORT.getAndIncrement();
            if (isBindable(port)) {
                return port;
            }
        }
        throw new IllegalStateException("failed to allocate tcp port");
    }

    private static boolean isBindable(int port) {
        try (ServerSocket server = new ServerSocket(port)) {
            server.setReuseAddress(false);
            return true;
        } catch (IOException ex) {
            return false;
        }
    }

    static void ensureNative() {
        Zlink.version();
    }

    static void await(CountDownLatch latch, String label) {
        try {
            if (!latch.await(TIMEOUT.toMillis(), TimeUnit.MILLISECONDS)) {
                throw new IllegalStateException(label + " timed out");
            }
        } catch (InterruptedException ex) {
            Thread.currentThread().interrupt();
            throw new IllegalStateException(label + " interrupted", ex);
        }
    }

    static void waitUntil(String label, Condition condition) {
        long deadlineNanos = System.nanoTime() + TIMEOUT.toNanos();
        while (System.nanoTime() < deadlineNanos) {
            if (condition.check()) {
                return;
            }
            sleepQuietly(label);
        }
        throw new IllegalStateException(label + " timed out");
    }

    private static void sleepQuietly(String label) {
        try {
            TimeUnit.MILLISECONDS.sleep(10);
        } catch (InterruptedException ex) {
            Thread.currentThread().interrupt();
            throw new IllegalStateException(label + " interrupted", ex);
        }
    }

    static String singleUtf8(Received received) {
        return received.singlePartOrThrow().toUtf8String();
    }

    static void waitConnected(SocketMonitor... monitors) {
        for (SocketMonitor monitor : monitors) {
            monitor.recv();
        }
    }

    static void waitStreamConnected(SocketMonitor monitor) {
        while (true) {
            var event = monitor.recv();
            if (event.event() == MonitorEventType.ACCEPTED
                || event.event() == MonitorEventType.CONNECTION_READY) {
                return;
            }
        }
    }

    static void waitPubSubReady(SocketMonitor pubMonitor,
                                SocketMonitor subMonitor) {
        waitMonitorEvent(subMonitor, MonitorEventType.CONNECTION_READY);
        waitMonitorEvent(pubMonitor, MonitorEventType.CONNECTION_READY);
    }

    /** Waits until the mesh node has admitted at least one peer. */
    static void waitMeshPeerConnected(MeshNode node) {
        waitUntil("mesh peer admission",
            () -> node.status().admittedPeerCount() > 0);
    }

    /**
     * Drains the node's ready index once and invokes {@code handler} for every
     * received record. The handler may take ownership of a record's parts with
     * {@code batch.retainMessage(index)} and answer request/control records via
     * {@link Dispatch}.
     */
    static void pumpReady(MeshNode node, ReadyBatch ready, ReceiveBatch recv,
                          RecordHandler handler) {
        pumpReady(node, ready, recv, handler, ReadyDomain.mask(ReadyDomain.ALL));
    }

    static void pumpReady(MeshNode node, ReadyBatch ready, ReceiveBatch recv,
                          RecordHandler handler, int domains) {
        ready.reset();
        node.drainReady(domains, ready, RecvFlags.DONT_WAIT);
        int readyCount = ready.count();
        for (int i = 0; i < readyCount; i++) {
            try (Claim claim = ready.takeClaim(i)) {
                while (claim.valid()) {
                    recv.reset();
                    ClaimRecvResult result = claim.recvBatch(recv,
                        RecvFlags.DONT_WAIT);
                    if (result.resultCode() != 0) {
                        break;
                    }
                    int count = recv.count();
                    for (int r = 0; r < count; r++) {
                        handler.handle(recv.at(r), recv, r);
                    }
                }
            }
        }
    }

    /**
     * Joins a local actor to a spot hosted on the same node: submits the join,
     * admits the host-side control record with {@code admitPayload}, waits for
     * the join completion, and returns the actor's membership epoch.
     */
    static long joinLocalSpot(MeshNode node, Actor actor, Spot spot,
                              String joinPayload, String admitPayload,
                              Consumer<String> onHostObservedJoin) {
        try (ReadyBatch ready = ReadyBatch.create(16);
             ReceiveBatch recv = ReceiveBatch.create(64, 256, 1 << 16)) {
            OperationId joinOp;
            try (Message join = Message.from(joinPayload)) {
                joinOp = actor.joinSpot(node.status().routingId(),
                    spot.routingId(), spot.status().lifecycleGeneration(),
                    List.of(join), TIMEOUT);
            }
            boolean[] completed = {false};
            int[] terminal = {-1};
            long deadline = System.nanoTime() + TIMEOUT.toNanos();
            while (!completed[0] && System.nanoTime() < deadline) {
                pumpReady(node, ready, recv, (record, batch, index) -> {
                    if (record.kind() == RecordKind.SPOT_CONTROL
                        && record.operationKind() == OperationKind.ACTOR_JOIN) {
                        List<Message> request = batch.retainMessage(index);
                        if (onHostObservedJoin != null && !request.isEmpty()) {
                            onHostObservedJoin.accept(
                                request.get(0).toUtf8String());
                        }
                        try (Message ok = Message.from(admitPayload)) {
                            Dispatch.actorJoinReply(record.replyToken(),
                                ActorJoinDecision.ACCEPTED, List.of(ok),
                                SendFlags.NONE);
                        }
                        closeAll(request);
                    } else if (record.kind() == RecordKind.COMPLETION
                        && record.operationKind() == OperationKind.ACTOR_JOIN
                        && record.operationId().equals(joinOp)) {
                        terminal[0] = record.terminalResult();
                        completed[0] = true;
                    }
                });
                if (completed[0]) {
                    break;
                }
                sleepQuietly("actor join");
            }
            if (!completed[0]) {
                throw new IllegalStateException("actor join did not complete");
            }
            if (terminal[0] != 0) {
                throw new IllegalStateException(
                    "actor join rejected: " + terminal[0]);
            }
            return node.actorLookup(actor.ref().actorId())
                .map(ActorLocation::membershipEpoch)
                .orElse(0L);
        }
    }

    /** Leaves a spot the actor previously joined (best effort completion wait). */
    static void leaveLocalSpot(MeshNode node, Actor actor, long membershipEpoch) {
        try (ReadyBatch ready = ReadyBatch.create(16);
             ReceiveBatch recv = ReceiveBatch.create(64, 256, 1 << 16)) {
            OperationId leaveOp = actor.leaveSpot(membershipEpoch, TIMEOUT);
            boolean[] completed = {false};
            long deadline = System.nanoTime() + TIMEOUT.toNanos();
            while (!completed[0] && System.nanoTime() < deadline) {
                pumpReady(node, ready, recv, (record, batch, index) -> {
                    if (record.kind() == RecordKind.COMPLETION
                        && record.operationKind() == OperationKind.ACTOR_LEAVE
                        && record.operationId().equals(leaveOp)) {
                        completed[0] = true;
                    }
                });
                if (completed[0]) {
                    break;
                }
                sleepQuietly("actor leave");
            }
        }
    }

    /** Creates and starts a STREAM session service for a stream socket. */
    static StreamSessionService startSessionService(MeshNode node,
                                                    StreamSocket stream) {
        StreamSessionService service = StreamSessionService.create(node, stream);
        service.start();
        return service;
    }

    /** Binds an actor to a STREAM session and waits for the bind completion. */
    static void bindSessionActor(MeshNode node, StreamSessionService service,
                                 RoutingId sessionRid, ActorRef actor) {
        OperationId op = service.bindActor(sessionRid, actor, TIMEOUT);
        waitStreamOperation(node, op, "bind");
    }

    /** Unbinds an actor from a STREAM session through the session service. */
    static void unbindSessionActor(MeshNode node, StreamSessionService service,
                                   RoutingId sessionRid, ActorRef actor) {
        long generation = 0;
        for (StreamSessionBinding binding : service.bindings(sessionRid)) {
            if (binding.actor().actorId().equals(actor.actorId())) {
                generation = binding.bindingGeneration();
            }
        }
        OperationId op = service.unbindActor(sessionRid, actor, generation,
            TIMEOUT);
        waitStreamOperation(node, op, "unbind");
    }

    /** Relays a payload to a session-bound actor via the session service. */
    static void relaySessionMessage(StreamSessionService service,
                                    RoutingId sessionRid, ActorRef actor,
                                    String payload) {
        try (Message message = Message.from(payload)) {
            service.sendToActor(sessionRid, actor, List.of(message),
                SendFlags.NONE);
        }
    }

    private static void waitStreamOperation(MeshNode node, OperationId opId,
                                            String name) {
        try (ReadyBatch ready = ReadyBatch.create(16);
             ReceiveBatch recv = ReceiveBatch.create(64, 256, 1 << 16)) {
            boolean[] completed = {false};
            int[] terminal = {0};
            long deadline = System.nanoTime() + TIMEOUT.toNanos();
            while (!completed[0] && System.nanoTime() < deadline) {
                pumpReady(node, ready, recv, (record, batch, index) -> {
                    if (record.kind() == RecordKind.COMPLETION
                        && record.operationId().equals(opId)) {
                        terminal[0] = record.terminalResult();
                        completed[0] = true;
                    }
                });
                if (completed[0]) {
                    break;
                }
                sleepQuietly("stream session " + name);
            }
            if (!completed[0]) {
                throw new IllegalStateException(
                    "stream session " + name + " did not complete");
            }
            if (terminal[0] != 0) {
                throw new IllegalStateException(
                    "stream session " + name + " failed: " + terminal[0]);
            }
        }
    }

    /**
     * Drains the node once and appends every actor-addressed message payload
     * into {@code sink} in arrival order.
     */
    static void collectActorMessages(MeshNode node, ReadyBatch ready,
                                     ReceiveBatch recv, List<String> sink) {
        pumpReady(node, ready, recv, (record, batch, index) -> {
            if (record.kind() != RecordKind.ACTOR_SEND
                || record.partCount() == 0) {
                return;
            }
            List<Message> parts = batch.retainMessage(index);
            if (!parts.isEmpty()) {
                sink.add(parts.get(0).toUtf8String());
            }
            closeAll(parts);
        });
    }

    static void closeAll(List<Message> parts) {
        for (Message part : parts) {
            closeQuietly(part);
        }
    }

    static void closeQuietly(AutoCloseable resource) {
        if (resource == null) {
            return;
        }
        try {
            resource.close();
        } catch (Exception ignored) {
        }
    }

    static java.net.Socket connectRawTcp(String endpoint) {
        try {
            InetSocketAddress address = tcpAddress(endpoint);
            java.net.Socket socket = new java.net.Socket();
            socket.connect(address, (int) TIMEOUT.toMillis());
            socket.setSoTimeout((int) TIMEOUT.toMillis());
            return socket;
        } catch (IOException ex) {
            throw new IllegalStateException("failed to connect raw tcp client", ex);
        }
    }

    static void sendRawTcp(java.net.Socket socket, byte[] payload) {
        try {
            OutputStream out = socket.getOutputStream();
            out.write(payload);
            out.flush();
        } catch (IOException ex) {
            throw new IllegalStateException("failed to send raw tcp payload", ex);
        }
    }

    /** Sends a STREAM socket packet frame (empty header, big-endian lengths). */
    static void sendStreamPacket(java.net.Socket socket, byte[] body) {
        byte[] frame = new byte[6 + body.length];
        frame[2] = (byte) (body.length >>> 24);
        frame[3] = (byte) (body.length >>> 16);
        frame[4] = (byte) (body.length >>> 8);
        frame[5] = (byte) body.length;
        System.arraycopy(body, 0, frame, 6, body.length);
        sendRawTcp(socket, frame);
    }

    static byte[] recvExactRawTcp(java.net.Socket socket, int length) {
        try {
            InputStream in = socket.getInputStream();
            return readExactly(in, length);
        } catch (IOException ex) {
            throw new IllegalStateException("failed to receive raw tcp payload", ex);
        }
    }

    private static InetSocketAddress tcpAddress(String endpoint) {
        if (!endpoint.startsWith("tcp://")) {
            throw new IllegalArgumentException(
                "raw tcp helper requires tcp:// endpoint: " + endpoint);
        }
        String hostPort = endpoint.substring("tcp://".length());
        int colon = hostPort.lastIndexOf(':');
        if (colon <= 0 || colon == hostPort.length() - 1) {
            throw new IllegalArgumentException("invalid tcp endpoint: " + endpoint);
        }
        String host = hostPort.substring(0, colon);
        int port = Integer.parseInt(hostPort.substring(colon + 1));
        return new InetSocketAddress(host, port);
    }

    private static byte[] readExactly(InputStream in, int length)
      throws IOException {
        byte[] data = new byte[length];
        int read = 0;
        while (read < length) {
            int n = in.read(data, read, length - read);
            if (n < 0) {
                throw new IOException("unexpected end of stream");
            }
            read += n;
        }
        return data;
    }

    private static void waitMonitorEvent(SocketMonitor monitor,
                                         MonitorEventType eventType) {
        while (true) {
            var event = monitor.recv();
            if (event.event() == eventType) {
                return;
            }
        }
    }

    @FunctionalInterface
    interface Condition {
        boolean check();
    }
}

File: bindings/dotnet/samples/SampleCommon/SampleSupport.cs
using System;
using System.Buffers.Binary;
using System.IO;
using System.Net;
using System.Net.Sockets;
using System.Text;
using System.Threading;
using Systems.Zlink;

namespace SampleCommon;

public static class SampleSupport
{
    public static bool IsNativeAvailable()
    {
        try
        {
            _ = global::Systems.Zlink.Zlink.Version();
            return true;
        }
        catch (DllNotFoundException)
        {
            return false;
        }
        catch (EntryPointNotFoundException)
        {
            return false;
        }
        catch (TypeInitializationException ex) when (ex.InnerException
                is DllNotFoundException or EntryPointNotFoundException)
        {
            return false;
        }
    }

    public static string NewEndpoint(string transport, string prefix)
    {
        int port = ReservePort();
        return $"{transport}://127.0.0.1:{port}";
    }

    public static RoutingId RoutingIdUtf8(string value)
    {
        return RoutingId.From(Encoding.UTF8.GetBytes(value));
    }

    public static int ReservePort()
    {
        TcpListener listener = new(IPAddress.Loopback, 0);
        listener.Start();
        int port = ((IPEndPoint)listener.LocalEndpoint).Port;
        listener.Stop();
        return port;
    }

    public static void WaitConnected(params ISocketMonitor[] monitors)
    {
        foreach (ISocketMonitor monitor in monitors)
            WaitMonitorEvent(monitor, 5000, SocketEvent.ConnectionReady);
    }

    public static MonitorEvent WaitMonitorEvent(ISocketMonitor monitor,
        int timeoutMs, params SocketEvent[] expectedEvents)
    {
        if (expectedEvents == null || expectedEvents.Length == 0)
        {
            throw new ArgumentException("Expected monitor events are required.",
                nameof(expectedEvents));
        }

        _ = timeoutMs;
        MonitorEvent evt = monitor.Recv()
            ?? throw new TimeoutException("Timed out waiting for monitor event.");
        for (int i = 0; i < expectedEvents.Length; i++)
        {
            if ((SocketEvent)evt.Event == expectedEvents[i])
                return evt;
        }

        throw new InvalidOperationException(
            $"Unexpected monitor event {evt.Event}.");
    }

    public static void WaitOrThrow(Func<bool> predicate, int timeoutMs,
        string message)
    {
        DateTime deadline = DateTime.UtcNow.AddMilliseconds(timeoutMs);
        while (DateTime.UtcNow < deadline)
        {
            if (predicate())
                return;
            Thread.Sleep(10);
        }

        throw new TimeoutException(message);
    }

    /// <summary>
    ///     Configures a MeshNode with the identity, bind endpoint, and channel
    ///     Core requires before <c>Start</c> (a unique routing id, a bind, and at
    ///     least one channel), then starts it.
    /// </summary>
    public static void StartConfiguredNode(IMeshNode node, string routingId,
        string channelName = "app", string bind = "tcp://127.0.0.1:*")
    {
        node.SetRoutingId(RoutingId.From(routingId));
        node.SetBind(bind);
        node.AddChannel(channelName);
        node.Start();
    }

    /// <summary>Creates and starts a STREAM session service for a stream socket.</summary>
    public static IStreamSessionService StartSessionService(IMeshNode node,
        IStreamSocket stream)
    {
        IStreamSessionService service = node.CreateStreamSessionService(stream);
        service.Start();
        return service;
    }

    /// <summary>
    ///     Binds an actor to a STREAM session through the session service and
    ///     waits for the bind completion on the node's ready index.
    /// </summary>
    public static void BindSessionActor(IMeshNode node,
        IStreamSessionService service, RoutingId sessionRid, ActorRef actor,
        int timeoutMs = 5000)
    {
        SubmitResult rc = service.BindActor(sessionRid, actor,
            out MeshOperationId op);
        if (rc != SubmitResult.Ok)
            throw new InvalidOperationException(
                $"stream session bind submit failed: {rc}");
        WaitStreamOperation(node, op, "bind", timeoutMs);
    }

    /// <summary>Unbinds an actor from a STREAM session through the session service.</summary>
    public static void UnbindSessionActor(IMeshNode node,
        IStreamSessionService service, RoutingId sessionRid, ActorRef actor,
        int timeoutMs = 5000)
    {
        ulong generation = 0;
        foreach (StreamSessionBinding binding in service.Bindings(sessionRid))
        {
            if (string.Equals(binding.Actor.ActorId, actor.ActorId,
                    StringComparison.Ordinal))
                generation = binding.BindingGeneration;
        }

        SubmitResult rc = service.UnbindActor(sessionRid, actor, generation,
            out MeshOperationId op);
        if (rc != SubmitResult.Ok)
            throw new InvalidOperationException(
                $"stream session unbind submit failed: {rc}");
        WaitStreamOperation(node, op, "unbind", timeoutMs);
    }

    /// <summary>Relays a payload to a session-bound actor via the session service.</summary>
    public static void RelaySessionMessage(IStreamSessionService service,
        RoutingId sessionRid, ActorRef actor, string payload)
    {
        using Message message = Message.From(payload);
        SubmitResult rc = service.SendToActor(sessionRid, actor,
            new[] { message });
        if (rc != SubmitResult.Ok)
            throw new InvalidOperationException(
                $"stream session relay failed: {rc}");
    }

    private static void WaitStreamOperation(IMeshNode node, MeshOperationId opId,
        string name, int timeoutMs)
    {
        using var ready = new MeshReadyBatch();
        using var recv = new MeshReceiveBatch();
        bool completed = false;
        int terminal = 0;
        DateTime deadline = DateTime.UtcNow.AddMilliseconds(timeoutMs);
        while (!completed && DateTime.UtcNow < deadline)
        {
            PumpReady(node, ready, recv, (record, batch, index) =>
            {
                if (record.Kind == MeshRecordKind.Completion
                    && record.OperationId.Equals(opId))
                {
                    terminal = record.TerminalResult;
                    completed = true;
                }
            });
            if (completed)
                break;
            Thread.Sleep(10);
        }

        if (!completed)
            throw new TimeoutException(
                $"stream session {name} did not complete");
        if (terminal != 0)
            throw new InvalidOperationException(
                $"stream session {name} failed: {terminal}");
    }

    public static void WaitSpotPeerConnected(IMeshNode node, int timeoutMs = 5000)
    {
        WaitOrThrow(
            () => node.Status().AdmittedPeerCount > 0,
            timeoutMs,
            "mesh peer admission");
    }

    /// <summary>
    ///     Drains the node's ready index once and invokes <paramref name="handler" />
    ///     for every received record. The handler may take ownership of a record's
    ///     parts with <c>batch.RetainMessage(index)</c> and answer request or control
    ///     records with <c>record.Reply(...)</c>.
    /// </summary>
    public static void PumpReady(IMeshNode node, MeshReadyBatch ready,
        MeshReceiveBatch recv,
        Action<MeshReceiveRecord, MeshReceiveBatch, int> handler,
        MeshReadyDomains domains = MeshReadyDomains.All)
    {
        ready.Reset();
        node.DrainReady(domains, ready, RecvFlags.DontWait);
        int readyCount = ready.Count;
        for (int i = 0; i < readyCount; i++)
        {
            using MeshClaim claim = ready.TakeClaim(i);
            recv.Reset();
            while (claim.Receive(recv, RecvFlags.DontWait))
            {
                int count = recv.Count;
                for (int r = 0; r < count; r++)
                    handler(recv[r], recv, r);
                recv.Reset();
            }
        }
    }

    /// <summary>
    ///     Joins a local actor to a spot hosted on the same node. Submits the join,
    ///     admits the incoming host-side control record with <paramref name="admitPayload" />,
    ///     waits for the join completion, and returns the actor's membership epoch
    ///     (needed later to leave). Optionally verifies the join payload the host
    ///     observed.
    /// </summary>
    public static ulong JoinLocalSpot(IMeshNode node, ActorRef actor, ISpot spot,
        string joinPayload, string admitPayload = "accepted",
        Action<string>? onHostObservedJoin = null, int timeoutMs = 5000)
    {
        using var ready = new MeshReadyBatch();
        using var recv = new MeshReceiveBatch();

        MeshOperationId joinOp;
        using (Message join = Message.From(joinPayload))
        {
            joinOp = node.JoinSpot(actor, node.RoutingId, spot.RoutingId,
                spot.Status().LifecycleGeneration, new[] { join });
        }

        bool completed = false;
        int terminal = -1;
        DateTime deadline = DateTime.UtcNow.AddMilliseconds(timeoutMs);
        while (!completed && DateTime.UtcNow < deadline)
        {
            PumpReady(node, ready, recv, (record, batch, index) =>
            {
                if (record.Kind == MeshRecordKind.SpotControl
                    && record.OperationKind == MeshOperationKind.ActorJoin)
                {
                    // 호스트가 join 요청 레코드를 받아 admit(응답)한다. actor join은
                    // 전용 join-reply 경로(record.Accept)로만 승인할 수 있다.
                    Message[] request = batch.RetainMessage(index);
                    if (onHostObservedJoin != null && request.Length > 0)
                        onHostObservedJoin(request[0].GetString());
                    using (Message ok = Message.From(admitPayload))
                        record.Accept(new[] { ok });
                    MultipartCloseSafe(request);
                }
                else if (record.Kind == MeshRecordKind.Completion
                    && record.OperationKind == MeshOperationKind.ActorJoin
                    && record.OperationId.Equals(joinOp))
                {
                    terminal = record.TerminalResult;
                    completed = true;
                }
            });
            if (completed)
                break;
            Thread.Sleep(10);
        }

        if (!completed)
            throw new TimeoutException("actor join did not complete");
        if (terminal != 0)
            throw new InvalidOperationException($"actor join rejected: {terminal}");

        return node.ActorLookup(actor.ActorId, out ActorLocation location)
            ? location.MembershipEpoch
            : 0UL;
    }

    /// <summary>
    ///     Leaves a spot the actor previously joined and waits (best effort) for the
    ///     leave completion.
    /// </summary>
    public static void LeaveLocalSpot(IMeshNode node, ActorRef actor,
        ulong membershipEpoch, int timeoutMs = 5000)
    {
        using var ready = new MeshReadyBatch();
        using var recv = new MeshReceiveBatch();
        MeshOperationId leaveOp = node.LeaveSpot(actor, membershipEpoch);

        bool completed = false;
        DateTime deadline = DateTime.UtcNow.AddMilliseconds(timeoutMs);
        while (!completed && DateTime.UtcNow < deadline)
        {
            PumpReady(node, ready, recv, (record, batch, index) =>
            {
                if (record.Kind == MeshRecordKind.Completion
                    && record.OperationKind == MeshOperationKind.ActorLeave
                    && record.OperationId.Equals(leaveOp))
                {
                    completed = true;
                }
            });
            if (completed)
                break;
            Thread.Sleep(10);
        }
    }

    /// <summary>
    ///     Drains the node once and appends every actor-addressed message payload
    ///     into <paramref name="sink" /> in arrival order.
    /// </summary>
    public static void CollectActorMessages(IMeshNode node, MeshReadyBatch ready,
        MeshReceiveBatch recv, List<string> sink)
    {
        PumpReady(node, ready, recv, (record, batch, index) =>
        {
            if (record.Kind != MeshRecordKind.ActorSend || record.PartCount == 0)
                return;
            Message[] parts = batch.RetainMessage(index);
            sink.Add(parts[0].GetString());
            MultipartCloseSafe(parts);
        });
    }

    private static void MultipartCloseSafe(Message[] parts)
    {
        if (parts.Length > 0)
            Zlink.MultipartClose(parts);
    }

    public static string ReceiveUtf8(IMessageSocket socket, int timeoutMs)
    {
        _ = timeoutMs;
        using var received = Received.Create();
        if (!socket.Recv(received))
            throw new InvalidOperationException("recv failed");
        if (received.Parts.Count == 0)
            throw new InvalidOperationException(
                "Expected at least one message part.");
        return Encoding.UTF8.GetString(received.Parts[0].AsReadOnlySpan());
    }

    public static string SubscribeUtf8(ISubscriberSocket socket, out string topic,
        int timeoutMs)
    {
        _ = timeoutMs;
        using var subscribed = new TopicMessage();
        socket.Subscribe(subscribed);
        topic = subscribed.Topic;
        if (subscribed.Parts.Count == 0)
            throw new InvalidOperationException(
                "Expected at least one subscribed message part.");
        return Encoding.UTF8.GetString(subscribed.Parts[0].AsReadOnlySpan());
    }

    public static TcpClient ConnectRawClient(int port)
    {
        var client = new TcpClient();
        client.NoDelay = true;
        client.Connect(IPAddress.Loopback, port);
        return client;
    }

    public static void SendAll(NetworkStream stream, ReadOnlySpan<byte> payload)
    {
        stream.Write(payload);
        stream.Flush();
    }

    public static void SendStreamPacket(NetworkStream stream,
        ReadOnlySpan<byte> body, ReadOnlySpan<byte> header = default)
    {
        if (header.Length > ushort.MaxValue)
        {
            throw new ArgumentOutOfRangeException(nameof(header),
                "Stream packet header is too large.");
        }

        byte[] frame = new byte[6 + header.Length + body.Length];
        BinaryPrimitives.WriteUInt16BigEndian(frame.AsSpan(0, 2),
            (ushort)header.Length);
        BinaryPrimitives.WriteUInt32BigEndian(frame.AsSpan(2, 4),
            (uint)body.Length);
        header.CopyTo(frame.AsSpan(6, header.Length));
        body.CopyTo(frame.AsSpan(6 + header.Length, body.Length));
        SendAll(stream, frame);
    }

    public static byte[] ReceiveExact(NetworkStream stream, int size)
    {
        byte[] buffer = new byte[size];
        int read = 0;
        while (read < size)
        {
            int n = stream.Read(buffer, read, size - read);
            if (n <= 0)
                throw new IOException("stream receive timeout");
            read += n;
        }

        return buffer;
    }

    public static int ExtractPort(string endpoint)
    {
        int idx = endpoint.LastIndexOf(':');
        return int.Parse(endpoint.AsSpan(idx + 1));
    }

    public static void EnsureEqual(string expected, string actual, string name)
    {
        if (!string.Equals(expected, actual, StringComparison.Ordinal))
        {
            throw new InvalidOperationException(
                $"Expected {name} \"{expected}\" but received \"{actual}\".");
        }
    }

}
